lib/using_restricted_array.rb

```diff
@@ -1,61 +1,117 @@
-require_relative 'restricted_array.rb'
+require_relative "restricted_array.rb"
 # RestrictedArray can be created using a specified size, or a random size in
 # the range of 1-20 will be chosen for you.
 # All values are integers in the range of 1-221.
 # RestrictedArray cannot be resized.
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
+# Time complexity: Linear or O(n), where n is the number of items in the array or the length of the input #array
 # Space complexity: ?
 def length(array)
-  raise NotImplementedError
+  i = 0
+  # until array[i] == nil
+  while array[i] != nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
+# Time complexity: linear or O(n),where n is the number of items in the array
 # Space complexity: ?
 def print_array(array)
-  raise NotImplementedError
+  # [3, 7, 9]
+  len = length(array)
+  i = 0
+  while i < len
+    print array[i]
+    print " "
+    i += 1
+  end
+  puts
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
+# Time complexity: O(n), where n is the number of values in the array
 # Space complexity: ?
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    return true if array[i] == value_to_find
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
+# Time complexity: O(n), where n is the number of values in the array
 # Space complexity: ?
 def find_largest(array, length)
-  raise NotImplementedError
+  largest = 0
+  i = 0
+  while i < length
+    if array[i] > largest
+      largest = array[i]
+    end
+    i += 1
+  end
+  return largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
+# Time complexity: O(n), where n is the number of values in the array
 # Space complexity: ?
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest = array[0]
+  i = 0
```

Comment:
One small detail about setting i=0 initially. The first time your while loop runs it will compare the first element of the array to itself, which is not necessary for this fn. If you set i=1, you can skip that first "self" comparison. 

The same goes for the `find_largest` function.